### ISSUE TITLE
STACK question not displaying variables correctly in different browsers

### ISSUE THREAD

**Issue by golenkovm** [maths/moodle-qtype_stack#809]
Reviewing some quiz attempts it's been found that some variables are not displaying correctly.

Expected/correct behaviour:
![image](https://user-images.githubusercontent.com/22066290/172082009-10b8ec47-8563-4439-978d-e005f6ec8a15.png)

Actual behaviour:
![image](https://user-images.githubusercontent.com/22066290/172082066-f342437e-1977-4b33-a6b0-dd9f6efce03f.png)
![image](https://user-images.githubusercontent.com/22066290/172081930-cf2ba875-709d-4898-b8b6-bf4b85848987.png)

Users were comlainging about one quiz in production site, but it worked well for me (and other admins). However, I was able to restore its backup and replicate the issue in test environment. Here is the outcome of my investigations:

There is a span element with class "MathJax_Preview". In test site it has a link in it like https://mymoodle.com/filter/tex/displaytex.php?texexp=%7B1.6%7D that returns GIF file. I noted that for some reason it can't be rendered by the browsers (when I open this link directly or check _response_ tab in Chome dev tools.

In the production site (where variables are displayed correctly) "MathJax_Preview" span is completely empty. Below this span there is another one with class "MathJax" (which can't be found in the test site) and this span holds data that seems to be responsible for correctly displayed numbers.

I also couldn't see `filter/tex/displaytex.php?texexp` requests (in _Network_ browser tab) in production site while could see them in test site.

There is another difference in HTML elements:
Test site:
• <script type="math/tex">{1.6}</script>
Production site:
• <script type="math/tex" id="MathJax-Element-1">{1.6}</script>

At the first glance it feels like there is some JS script that is supposed to process "MathJax_Preview" span, clean it up and create new span "MathJax" with correct data in it.

Has anyone faced this issue before?

Kind regards,
Misha

**Comment by aharjula** [comment]
Sounds like that Moodle has the TeX filter active and not that actual MathJax filter or maybe they are both there and are fighting for the chance to do something. The one higher in the filter list will win. MathJax is basically a script that does the rendering and the filter exists to mark the things it should target, if some other filter robs those targets it just won't work

I base this diagnosis on that URL that includes `/filter/tex/displaytex.php`. Seeing that in the output is a good indicator of that filter being active. Do compare the filter settings of those Moodles to see if this is the case.

Generally, I recommend only using the MathJax filter and if that is not an option only use one filter that affects TeX. The general filter guidance is to place localisation filters (if any are in use) at the top of the filter list and MathJax as the last one. Emoticons are the exception if those are in play (you probably don't want them) MathJax should be before them, even then they may break formulae.

For others reading this: You will need site admin rights on the Moodle to tune the filter order, when you have those rights you can at the same time remove pointless filters. Typically leave this task to the admins and once it is fixed you can forget it.

**Comment by aharjula** [comment]
If that is not the case, do check inside the question and see what actually reads on that line at that position that is not getting rendered. There is historical precedence, of an author copy-pasting output from some other system into a question so that even links to pictures (and other filtering results), were copied as links to pictures leading to just those singular items not displaying correctly and a long process of tracking the reason down. In general, you should not see any links to filters or those `<script type="math/tex"...`-wrappers in the question source code. In the case of STACK there should only be something in the style of `{@...@}` at that position.

Basically, this could be a reason for cases where the links point to the other Moodle and only function for those users that are logged in in both Moodles and have enough rights in both of them. It is also the reason why I really dislike all hard links to anything that is outside the question, as it tends to break so that authors and admins do not have any problems but everyone else has.

**Comment by golenkovm** [comment]
Thank you for the advice @aharjula 

Correct, both `MathJax` and `TeX notation` filter were active and `TeX notation` was on top of the filter list.
I can confirm that when I turn `TeX notation` off or move it lower than `MathJax` - the variables are displayed correctly.
When I put `TeX notation` back on top of the list - they are displayed as _/filter/tex/displaytex.php?texexp=..._ links again.

I'm wondering if this mentioned somewhere? I checked https://docs.stack-assessment.org/en/Installation/ but couldn't find that `TeX notation` needs to be disabled or moved lowver `MathJax`.

**Comment by golenkovm** [comment]
Sorry, all the tests described above were taken in the test site.

I checked settings in production site and they seem to have identical settings. At least filters go in the same order on the _/admin/filters.php_ page and also `TeX notation`/`MathJax` settings are in the same state and seem to be never changed.

This makes me wondering why in production site I can't replicate the issue.

**Comment by aharjula** [comment]
Note that one can tune which filters are active also at the course and quiz/activity levels. And I think even on (course) category level can tune that. So you may need to check the whole chain and not just the top-level site list of filters. Although I believe the filter setting at the end of the chain will display the actual setting at that level so you see the differences from that end if any exist. It could be that someone has fixed this in the production set-up at some of these levels and materials outside that "category" (or any of those others) are the only ones suffering from this.

If this is not mentioned in STACK documentation I would imagine it is something that the MathJax filter documents. But maybe you could contribute an addition to the docs as you seem to know the troubles a misconfiguration can cause?

**Comment by sangwinc** [comment]
If this is not in the STACK documentation it could also be that we have no experience of trying to use both MathJax and tex filters in parallel!  If you could track down what is going on with this combination I'll very happily document any problems (and the fix!) in the STACK docs.

**Comment by golenkovm** [comment]
You were right @aharjula, `TeX notation` was disabled on the course level, so this why I wasn't able to replicate the issue in Production siеe, but was able to restore the course in the test site and catch it there. Perfect, I really appreciate your help.

Hi @sangwinc, for some reason I thought that `TeX notation` is enabled by default. Checked vanilla Moodle settings and it's disabled. All good, no claims for stack documentation :+1: 

Thanks again @aharjula, @sangwinc. Happy to close the issue as resolved.

Cheers,
Misha
